### DISCUSSION TITLE
Tuya TS0041 not picking up custom quirk

### DISCUSSION THREAD

**Discussion by MrEcosse** [zigpy/zha-device-handlers#1731] [Q&A]
I appear to have three different 'manufacturers' of a TS0041 1 button keypad in my system.  Two are using the custom quirk fine in HA but the third is not using it.

Both the WORKING ones have the following signature:
`{
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.EndDevice: 2>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.AllocateAddress: 128>, manufacturer_code=4098, maximum_buffer_size=82, maximum_incoming_transfer_size=82, server_mask=11264, maximum_outgoing_transfer_size=82, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=True, *is_full_function_device=False, *is_mains_powered=False, *is_receiver_on_when_idle=False, *is_router=False, *is_security_capable=False)",
  "endpoints": {
    "1": {
      "profile_id": 260,
      "device_type": "0x0006",
      "in_clusters": [
        "0x0000",
        "0x0001",
        "0x0006"
      ],
      "out_clusters": [
        "0x000a",
        "0x0019"
      ]
    }
  },
  "manufacturer": "_TZ3000_tk3s5tyg",
  "model": "TS0041",
  "class": "zhaquirks.tuya.ts0041.TuyaSmartRemote0041TO"
}`

The not working one has the following signature
`{
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.EndDevice: 2>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.AllocateAddress: 128>, manufacturer_code=4098, maximum_buffer_size=82, maximum_incoming_transfer_size=82, server_mask=11264, maximum_outgoing_transfer_size=82, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=True, *is_full_function_device=False, *is_mains_powered=False, *is_receiver_on_when_idle=False, *is_router=False, *is_security_capable=False)",
  "endpoints": {
    "1": {
      "profile_id": 260,
      "device_type": "0x0000",
      "in_clusters": [
        "0x0000",
        "0x0001",
        "0x0006"
      ],
      "out_clusters": [
        "0x0019"
      ]
    }
  },
  "manufacturer": "_TYZB02_keyjqthh",
  "model": "TS0041",
  "class": "zigpy.device.Device"
}`

The differences are the device_type as 0x0000 and no out_cluster 0x000a.  I am not sure how or if this can be added to be recognised.

**Comment by MattWestb** [comment]
First redo the formatting by adding  ` ``` `  on one blank line before and after the code so the formatting is not broken.

You working device is using the TO (Time as Out cluster) and your problem device is not having the time cluster (suggested device class **TuyaSmartRemote0041WOT**).

Configure local quirks in HA and putting one copy of the https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/tuya/ts0041.py in the configured quirk folder.
Then deleting this 2 lines:
https://github.com/zigpy/zha-device-handlers/blob/b98948e7b6538ab31d87330bba1793e0eb65de11/zhaquirks/tuya/ts0041.py#L79
and https://github.com/zigpy/zha-device-handlers/blob/b98948e7b6538ab31d87330bba1793e0eb65de11/zhaquirks/tuya/ts0041.py#L94.
without braking the formatting.
Restart HA and look if the system is loading the quirk for your device.

**Reply by MattWestb** [reply]
The device type for both (more correct all TS0041) device is before the quirk `0x0000` and the quirk is changing it to `0x0006` for matching the function of the device in ZHA GUI then tuya have using the wrong device type in the firmware.

**Reply by MrEcosse** [reply]
The device is now picking up the quirk, but the Time as In Cluster - so sadly not showing the button press trigger events.  Have subsequently found a 3-button switch with the same behavior grr.
`
{
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.EndDevice: 2>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.AllocateAddress: 128>, manufacturer_code=4098, maximum_buffer_size=82, maximum_incoming_transfer_size=82, server_mask=11264, maximum_outgoing_transfer_size=82, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=True, *is_full_function_device=False, *is_mains_powered=False, *is_receiver_on_when_idle=False, *is_router=False, *is_security_capable=False)",
  "endpoints": {
    "1": {
      "profile_id": 260,
      "device_type": "0x0006",
      "in_clusters": [
        "0x0000",
        "0x0001",
        "0x0006"
      ],
      "out_clusters": [
        "0x0019"
      ]
    }
  },
  "manufacturer": "_TYZB02_keyjqthh",
  "model": "TS0041",
  "class": "ts0041.TuyaSmartRemote0041TI"
}
`
